File: src/data/codes/q-learning.m
clc;
clear;
close all;

%% =========================
%  GRID WORLD SETTINGS
% ==========================
rows = 4;
cols = 4;
numStates = rows * cols;
numActions = 4;   % 1=Up, 2=Down, 3=Left, 4=Right

startState = 1;       % top-left
goalState = 16;       % bottom-right
trapStates = [6 11];  % example trap cells

alpha = 0.1;          % learning rate
gamma = 0.9;          % discount factor
epsilon = 0.2;        % exploration rate
episodes = 500;
maxSteps = 50;

Q = zeros(numStates, numActions);

episodeRewards = zeros(episodes,1);
episodeSteps = zeros(episodes,1);

%% =========================
%  TRAINING
% ==========================
for ep = 1:episodes
    state = startState;
    totalReward = 0;

    for step = 1:maxSteps

        % Epsilon-greedy action selection
        if rand < epsilon
            action = randi(numActions);
        else
            [~, action] = max(Q(state,:));
        end

        % Take action and get next state
        nextState = getNextState(state, action, rows, cols);

        % Reward structure
        if nextState == goalState
            reward = 10;
        elseif ismember(nextState, trapStates)
            reward = -5;
        else
            reward = 0;
        end

        % Q-learning update
        Q(state, action) = Q(state, action) + alpha * ...
            (reward + gamma * max(Q(nextState,:)) - Q(state, action));

        state = nextState;
        totalReward = totalReward + reward;

        % Stop if goal reached
        if state == goalState
            break;
        end
    end

    episodeRewards(ep) = totalReward;
    episodeSteps(ep) = step;
end

%% =========================
%  DISPLAY Q-TABLE
% ==========================
disp('Learned Q-Table:');
disp(Q);

%% =========================
%  LEARNED POLICY
% ==========================
policy = strings(numStates,1);
actionsText = ["Up","Down","Left","Right"];

for s = 1:numStates
    if s == goalState
        policy(s) = "Goal";
    elseif ismember(s, trapStates)
        policy(s) = "Trap";
    else
        [~, bestAction] = max(Q(s,:));
        policy(s) = actionsText(bestAction);
    end
end

disp('Learned Policy:');
for r = 1:rows
    for c = 1:cols
        s = (r-1)*cols + c;
        fprintf('%8s ', policy(s));
    end
    fprintf('\n');
end

%% =========================
%  EVALUATION METRICS
% ==========================
avgReward = mean(episodeRewards);
avgSteps = mean(episodeSteps);

fprintf('\nEvaluation Metrics:\n');
fprintf('Average Reward per Episode = %.4f\n', avgReward);
fprintf('Average Steps per Episode  = %.4f\n', avgSteps);

%% =========================
%  VISUALIZATION
% ==========================
figure;
plot(episodeRewards, 'LineWidth', 1.5);
xlabel('Episode');
ylabel('Total Reward');
title('Reward per Episode');
grid on;

figure;
plot(episodeSteps, 'LineWidth', 1.5);
xlabel('Episode');
ylabel('Steps');
title('Steps per Episode');
grid on;

%% =========================
%  TEST THE LEARNED PATH
% ==========================
disp('Learned Path from Start to Goal:');
state = startState;
path = state;

for k = 1:20
    if state == goalState
        break;
    end
    [~, action] = max(Q(state,:));
    nextState = getNextState(state, action, rows, cols);
    path(end+1) = nextState; %#ok<SAGROW>
    state = nextState;

    if ismember(state, trapStates)
        break;
    end
end

disp(path);

%% =========================
%  FUNCTION
% ==========================
function nextState = getNextState(state, action, rows, cols)
    [r, c] = ind2sub([rows cols], state);

    if action == 1      % Up
        r = max(r-1, 1);
    elseif action == 2  % Down
        r = min(r+1, rows);
    elseif action == 3  % Left
        c = max(c-1, 1);
    elseif action == 4  % Right
        c = min(c+1, cols);
    end

    nextState = sub2ind([rows cols], r, c);
end

File: src/data/codes/q-learning.py
"""
Q-learning on a grid world
# pip install numpy matplotlib
"""
import numpy as np
import matplotlib.pyplot as plt

ROWS, COLS = 4, 4
NUM_STATES = ROWS * COLS
NUM_ACTIONS = 4  # 0=Up, 1=Down, 2=Left, 3=Right
START_STATE = 0
GOAL_STATE = 15
TRAP_STATES = {5, 10}

alpha = 0.1
gamma = 0.9
epsilon = 0.2
N_EPISODES = 500
MAX_STEPS = 50

RNG = np.random.default_rng(42)


def idx_to_rc(s: int):
    r, c = divmod(s, COLS)
    return r, c


def rc_to_idx(r: int, c: int) -> int:
    return r * COLS + c


def get_next_state(state: int, action: int) -> int:
    r, c = idx_to_rc(state)
    if action == 0:
        r = max(r - 1, 0)
    elif action == 1:
        r = min(r + 1, ROWS - 1)
    elif action == 2:
        c = max(c - 1, 0)
    elif action == 3:
        c = min(c + 1, COLS - 1)
    return rc_to_idx(r, c)


Q = np.zeros((NUM_STATES, NUM_ACTIONS))
episode_rewards = np.zeros(N_EPISODES)
episode_steps = np.zeros(N_EPISODES, dtype=int)

for ep in range(N_EPISODES):
    state = START_STATE
    total_reward = 0.0
    for step in range(MAX_STEPS):
        if RNG.random() < epsilon:
            action = int(RNG.integers(0, NUM_ACTIONS))
        else:
            action = int(np.argmax(Q[state]))

        next_state = get_next_state(state, action)
        if next_state == GOAL_STATE:
            reward = 10.0
        elif next_state in TRAP_STATES:
            reward = -5.0
        else:
            reward = 0.0

        td_target = reward + gamma * np.max(Q[next_state])
        Q[state, action] += alpha * (td_target - Q[state, action])
        state = next_state
        total_reward += reward
        if state == GOAL_STATE:
            break
    episode_rewards[ep] = total_reward
    episode_steps[ep] = step + 1

print("Learned Q:\n", Q)

labels = ["Up", "Down", "Left", "Right"]
policy = []
for s in range(NUM_STATES):
    if s == GOAL_STATE:
        policy.append("Goal")
    elif s in TRAP_STATES:
        policy.append("Trap")
    else:
        policy.append(labels[int(np.argmax(Q[s]))])

print("Learned Policy:")
for r in range(ROWS):
    row = [policy[rc_to_idx(r, c)].ljust(8) for c in range(COLS)]
    print(" ".join(row))

print(f"\nAvg Reward/ep = {episode_rewards.mean():.4f}")
print(f"Avg Steps/ep  = {episode_steps.mean():.4f}")

fig, ax = plt.subplots()
ax.plot(episode_rewards, linewidth=1.5)
ax.set_xlabel("Episode")
ax.set_ylabel("Total Reward")
ax.set_title("Reward per Episode")
ax.grid(True)

fig2, ax2 = plt.subplots()
ax2.plot(episode_steps, linewidth=1.5)
ax2.set_xlabel("Episode")
ax2.set_ylabel("Steps")
ax2.set_title("Steps per Episode")
ax2.grid(True)

path = [START_STATE]
state = START_STATE
for _ in range(20):
    if state == GOAL_STATE:
        break
    action = int(np.argmax(Q[state]))
    state = get_next_state(state, action)
    path.append(state)
    if state in TRAP_STATES:
        break
print("Learned path:", path)
plt.show()
